Specify bytes-per-inode when creating image

jmstover opened the pull request:
When using image.create to make an overlay image, this will allow
the specifying of the size of an inode.

mke2fs default is 16K per inode, which if you have a lot of small
files, means you would hit the inode limit before you would hit
the space limit.

This PR adds a `-i` option to specify the size in bytes of the
inodes to mkfs.ext3.

For example:
```
$ singularity image.create -i 8192 /tmp/overlay.img
```

Issue: #1759

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

jmstover committed:
Specify bytes-per-inode when creating image

When using image.create to make an overlay image, this will allow
the specifying of the size of an inode.

mke2fs default is 16K per inode, which if you have a lot of small
files, means you would hit the inode limit before you would hit
the space limit.

This PR adds a `-i` option to specify the size in bytes of the
inodes to mkfs.ext3.

For example:
```
$ singularity image.create -i 8192 /tmp/overlay.img
```

Issue: #1759

GodloveD commented on libexec/cli/image.create.exec:
```diff
@@ -96,6 +102,19 @@ if [ -f "$SINGULARITY_IMAGE" ]; then
+# We want to exit before the 'dd' command
+INODE=""
+if [ -n "${SINGULARITY_INODESIZE:-}" ]; then
+    INODE="-i ${SINGULARITY_INODESIZE:-16384}"
```
Is the `:-16384}"` part of this really necessary?  We just checked in the previous line to make sure that `SINGULARITY_INODESIZE` is not emtpy.

jmstover replied:
No ... Call it me being psychotic an putting a default value in case it's empty even though I just checked that it wasn't empty...

GodloveD commented on libexec/cli/image.create.exec:
```diff
@@ -96,6 +102,19 @@ if [ -f "$SINGULARITY_IMAGE" ]; then
+
+    if [ $(expr ${SINGULARITY_INODESIZE} % 1024) -ne 0 ]; then
+        message ERROR "Given inode size of ${SINGULARITY_INODESIZE} is not divisible by 1024\n"
+        exit 1
```
Should this be an `ABORT 255`?

jmstover replied:
Sure... I can change that. Call it habit for a `exit 1` failure ...

GodloveD reviewed (changes requested):
I've got a few questions about the code inlined.  Also, can you add this to the `CHANGELOG.md` and to the `etc/bash_completion.d/singularity` file please?

jmstover committed:
@GodloveD requests

Don't put a default value when we just checked that it wasn't
empty... don't really need to do that.

Use ABORT to return from script instead of `exit 1`.

jmstover committed:
Add CHANGELOG.md and bash_completion entry

GodloveD commented on libexec/cli/image.create.info:
```diff
@@ -11,11 +11,14 @@ CREATE OPTIONS:
     -s|--size   Specify a size for an operation in MiB, i.e. 1024*1024B
                 (default 768MiB)
     -F|--force  Overwrite an image file if it exists
+    -i          Specify a size for the inode ratio in Bytes. Default is
```
Oh sorry one more thing.  I think it's OK to have a long option without a short option, but I don't think we should have a short option without a long option.

GodloveD reviewed (dismissed):
Sorry one more thing.

jmstover committed:
Help info

Add CLI help info, and bash completion for the new option

jmstover committed:
Cleanup completion on --size

Don't call _filedir on -s|--size option. It's going to do a dir
listing if you hit tab again anyway, and we're a numric option.

jmstover committed:
Merge branch 'development-2.x' into mkfs_inode

GodloveD commented on etc/bash_completion.d/singularity:
```diff
@@ -369,7 +371,7 @@ _singularity() {
 
                 create)
```
Also needs the `-i` option to be added here:

```
11:13 AM
vagrant@ubuntu-xenial:~/singularity
$ singularity image create --
--force  --help   --size

11:13 AM
vagrant@ubuntu-xenial:~/singularity
$ singularity image.create --
--bytes-per-inode  --force            --help             --size
```

GodloveD commented on libexec/cli/image.create.exec:
```diff
@@ -96,6 +102,19 @@ if [ -f "$SINGULARITY_IMAGE" ]; then
+if [ -n "${SINGULARITY_INODESIZE:-}" ]; then
+    INODE="-i ${SINGULARITY_INODESIZE}"
+
+    if [ $(expr ${SINGULARITY_INODESIZE} % 1024) -ne 0 ]; then
```
This check is unnecessary.  We are not specifying the size of inodes, but the number of inodes in relation to the size of the disk space.  

On my system there is a minimum and maximum ratio that can be specified.  (min 1024/max 67108864), but I don't know how system specific so I'm hesitant to say we should check that the value is between that min and max.  

If you specify a value outside of that range, it will issue a warning and then create a broken file system for you.  I think the warning is pretty good though, and I'm inclined to just let it go at that.  

```
$ singularity image.create --bytes-per-inode 256 --size 512 ~/blah2
Creating empty 512MiB image file: /home/vagrant/blah2
Formatting image with ext3 file system
mkfs.ext3: invalid inode ratio 256 (min 1024/max 67108864)
```

GodloveD reviewed (changes requested):
Still needs a little more love I think.

jmstover committed:
Updated changes

- Add bash_completion to `image create` command
- Make the bytes-per-inode check a WARN instead of ERROR.

GodloveD commented on etc/bash_completion.d/singularity:
```diff
@@ -369,9 +371,11 @@ _singularity() {
                     if [[ ${cur} == -s || ${cur} == --size ]]; then
-                        _filedir
+                        return 0
+                    if [[ ${cur} == -i || ${cur} == --bytes-per-inode ]]; then
```
Thars a syntax error here.  Needs to be an `elif`

jmstover replied:
Why you care about syntax errors? It should **know** what I _mean_ ...

jmstover committed:
else if ... *else* if ...

GodloveD reviewed (approved):
lgtm!

ArangoGutierrez reviewed (approved):
LGTM

jmstover committed:
Merge branch 'development-2.x' into mkfs_inode
